File: src/clipboard.js
let copyBtns = document.querySelectorAll(".copy-btn");

async function copyContent(textToCopy, copyIcon, successIcon, errorIcon) {
    try {
        await navigator.clipboard.writeText(textToCopy);
        copyIcon.classList.add("hidden");
        successIcon.classList.remove("hidden");
        setTimeout(() => {
            copyIcon.classList.remove("hidden");
            successIcon.classList.add("hidden");
        }, 1000);
    } catch (err) {
        copyIcon.classList.add("hidden");
        errorIcon.classList.remove("hidden");
        setTimeout(() => {
            copyIcon.classList.remove("hidden");
            errorIcon.classList.add("hidden");
        }, 1000);
    }
}

copyBtns.forEach((copyBtn) => {
    copyBtn.addEventListener("click", () => {
        const container = copyBtn.closest("div");
        const textElement = container.querySelector(".font-mono");
        const textToCopy = textElement.innerText;

        copyContent(
            textToCopy,
            copyBtn.querySelector(".copy"),
            copyBtn.querySelector(".success"),
            copyBtn.querySelector(".failed")
        );
    });
});

File: WebSharperWebsite/Src/Clipboard.fs
namespace WebSharperWebsite

open WebSharper
open WebSharper.JavaScript
open WebSharper.JavaScript.Dom
open WebSharper.UI
open WebSharper.UI.Templating

[<JavaScript>]
module Clipboard =

    [<Inline "navigator.clipboard.writeText($0)">]
    let clipboardWriteText (v: string) = X<Promise<unit>> 

    let private show (el: Element) = 
        if not (isNull el) then 
            el.ClassList.Remove("hidden")

    let private hide (el: Element) = 
        if not (isNull el) then 
            el.ClassList.Add("hidden")

    let private revertAfter (copyIcon: Element) (otherIcon: Element) (ms: int) =
        JS.SetTimeout(fun () -> 
            show copyIcon
            hide otherIcon
        ) ms |> ignore

    let private copyContentAsync (textToCopy: string) (copyIcon: Element) (okIcon: Element) (errIcon: Element) = async {
        try
            do! Promise.AsAsync <| clipboardWriteText textToCopy
            hide copyIcon
            show okIcon
            revertAfter copyIcon okIcon 1000
        with
        | _ ->
            hide copyIcon
            show errIcon
            revertAfter copyIcon errIcon 1000
    }

    let CopyFromClosest (e: Runtime.Server.TemplateEvent<_, _, _>) =
        let button = e.Target
        let btn =
            match e.Event.Target with
            | :? Element as el -> el
            | _ -> JS.Document.ActiveElement |> As<Element>

        let textElement = JS.Document.QuerySelector(".text-to-copy")
        let textToCopy = textElement.TextContent

        let copyIcon = button.QuerySelector(".copy")
        let successIcon = button.QuerySelector(".success")
        let errorIcon = button.QuerySelector(".failed")

        copyContentAsync textToCopy copyIcon successIcon errorIcon
        |> Async.StartImmediate 

